src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/math/RobustBrentSolver.java

```diff
@@ -0,0 +1,439 @@
+
+            final double t = getAbsoluteAccuracy();
+            final double eps = getRelativeAccuracy();
+
+            while (true) {
+                if (FastMath.abs(fc) < FastMath.abs(fb)) {
+                    a = b;
+                    b = c;
+                    c = a;
+                    fa = fb;
+                    fb = fc;
+                    fc = fa;
+                }
+
+                final double tol = 2 * eps * FastMath.abs(b) + t;
+                final double m = 0.5 * (c - b);
+
+                if (FastMath.abs(m) <= tol ||
+                        Precision.equals(fb, 0))  {
+                    return b;
+                }
+                if (FastMath.abs(e) < tol ||
+                        FastMath.abs(fa) <= FastMath.abs(fb)) {
+                    // Force bisection.
+                    d = m;
+                    e = d;
+                } else {
+                    double s = fb / fa;
+                    double p;
+                    double q;
+                    // The equality test (a == c) is intentional,
+                    // it is part of the original Brent's method and
+                    // it should NOT be replaced by proximity test.
+                    if (a == c) {
+                        // Linear interpolation.
+                        p = 2 * m * s;
+                        q = 1 - s;
+                    } else {
+                        // Inverse quadratic interpolation.
+                        q = fa / fc;
+                        final double r = fb / fc;
+                        p = s * (2 * m * q * (q - r) - (b - a) * (r - 1));
+                        q = (q - 1) * (r - 1) * (s - 1);
+                    }
+                    if (p > 0) {
+                        q = -q;
+                    } else {
+                        p = -p;
+                    }
+                    s = e;
+                    e = d;
+                    if (p >= 1.5 * m * q - FastMath.abs(tol * q) ||
+                            p >= FastMath.abs(0.5 * s * q)) {
+                        // Inverse quadratic interpolation gives a value
+                        // in the wrong direction, or progress is slow.
+                        // Fall back to bisection.
+                        d = m;
+                        e = d;
+                    } else {
+                        d = p / q;
+                    }
+                }
+                a = b;
+                fa = fb;
+
+                if (FastMath.abs(d) > tol) {
+                    b += d;
+                } else if (m > 0) {
+                    b += tol;
+                } else {
+                    b -= tol;
+                }
+                fb = computeObjectiveValue(b);
+                if ((fb > 0 && fc > 0) ||
+                        (fb <= 0 && fc <= 0)) {
+                    c = a;
+                    fc = fa;
+                    d = b - a;
+                    e = d;
+                }
+            }
+        }
+    }
+}
```

Comment:
Something like the following (I've removed argument validation and the grid methods):

````
public final class RobustBrentSolver extends AbstractUnivariateSolver {
    private final UnivariateFunction meritFunc;
    private final int numBisections;
    private final int depth;

    public RobustBrentSolver(final double relativeAccuracy, 
                             final double absoluteAccuracy, 
                             final double functionValueAccuracy, 
                             final UnivariateFunction meritFunc,
                             final int numBisections, 
                             final int depth) {
        super(relativeAccuracy, absoluteAccuracy, functionValueAccuracy);
        this.numBisections = numBisections;
        this.depth = depth;
        this.meritFunc = meritFunc;
    }

    @Override
    public double solve(final int maxEval,
                        final UnivariateFunction objFunc,
                        final double min, 
                        final double max)
            throws TooManyEvaluationsException, NoBracketingException {
        setup(maxEval, objFunc, min, max, min);   //last parameter is not actually used
        return doSolve();
    }

    @Override
    protected double doSolve() throws TooManyEvaluationsException, NoBracketingException {
        final double min = getMin();
        final double max = getMax();
        final double[] xSearchGrid = createFractalSearchGrid(min, max, numBisections, depth);
        final double[] fSearchGrid = new double[xSearchGrid.length];
        for (int k = 0; k < xSearchGrid.length; k++) {
            fSearchGrid[k] = computeObjectiveValue(xSearchGrid[k]);
        }
        final List<Bracket> bracketsList = detectBrackets(xSearchGrid, fSearchGrid);
        if (bracketsList.isEmpty()) {
            throw new NoBracketingException(min, max, fSearchGrid[0], fSearchGrid[fSearchGrid.length-1]);
        }
        final List<Double> roots = new ArrayList<>(bracketsList.size());
        for (final Bracket bracket : bracketsList) {
            final double x0 = 0.5 * (bracket.xMin + bracket.xMax);
            final BrentSolver solver = new BrentSolver(
                    getRelativeAccuracy(), getAbsoluteAccuracy(), getFunctionValueAccuracy());
            final double x = solver.solve(getMaxEvaluations(), this::computeObjectiveValue, bracket.xMin, bracket.xMax, x0);
            roots.add(x);
        }
        if (roots.isEmpty()) {
            throw new NoBracketingException(min, max, fSearchGrid[0], fSearchGrid[fSearchGrid.length-1]);
        }
        if (roots.size() == 1) {
            return roots.get(0);
        }
        return roots.stream().max((r, s) -> Double.compare(meritFunc.value(r), meritFunc.value(s))).get();
    }
}